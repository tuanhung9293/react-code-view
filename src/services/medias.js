import camelize from 'camelize'

export const convertMedia = (medias) => {
  const camel = camelize(medias)
  const searchResults = camel.collection.items.map((item) => {
    return {href: item.href, ...item.data[0]}
  })
  return {
    data: searchResults,
    totalHits: camel.collection.metadata.totalHits
  }
}

export const getThumbUrl = (source) => {
  return source.find(item => item.includes('~small_thumb_') || item.includes('~thumb.'))
}

export const getMediaUrl = (source) => {
  return source.find(item => item.includes('~orig.') || item.includes('~small.'))
}