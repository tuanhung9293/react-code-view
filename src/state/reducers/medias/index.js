import {
  SEARCH_NASA_MEDIA,
  SEARCH_MORE_NASA_MEDIA,
  FETCH_MEDIA_ITEM_COLLECTION
} from 'constants/actionTypes'

const initialState = {
  data: [],
  totalHits: 0,
  params: {},
  loading: false,
  loadingMore: false
}

function medias(state = initialState, action) {
  switch (action.type) {
    case SEARCH_NASA_MEDIA.PENDING:
      return {
        ...state,
        loading: true
      }

    case SEARCH_MORE_NASA_MEDIA.PENDING:
      return {
        ...state,
        loadingMore: true
      }

    case SEARCH_NASA_MEDIA.SUCCESS:
      return {
        ...state,
        data: action.payload.data,
        totalHits: action.payload.totalHits,
        params: action.payload.params,
        loading: false
      }

    case SEARCH_MORE_NASA_MEDIA.SUCCESS:
      return {
        ...state,
        data: [...state.data, ...action.payload.data],
        params: action.payload.params,
        loadingMore: false
      }

    case FETCH_MEDIA_ITEM_COLLECTION.SUCCESS:
      let newStateData = [...state.data]
      newStateData[action.meta.index] = {
        ...newStateData[action.meta.index],
        thumbUrl: action.payload.thumbUrl,
        mediaUrl: action.payload.mediaUrl
      }

      return {
        ...state,
        data: newStateData
      }
    default:
      return state
  }
}

export default medias