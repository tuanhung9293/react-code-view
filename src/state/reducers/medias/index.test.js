import medias from './'
import {
  SEARCH_NASA_MEDIA,
  SEARCH_MORE_NASA_MEDIA,
  FETCH_MEDIA_ITEM_COLLECTION
} from 'constants/actionTypes'

describe('medias reducer', () => {
  it('sets initial state', () => {
    expect(medias(undefined, {})).toMatchSnapshot()
  })

  it('SEARCH_NASA_MEDIA.PENDING', () => {
    expect(medias(undefined, {type: SEARCH_NASA_MEDIA.PENDING})).toMatchSnapshot()
  })

  it('SEARCH_MORE_NASA_MEDIA.PENDING', () => {
    expect(medias(undefined, {type: SEARCH_MORE_NASA_MEDIA.PENDING})).toMatchSnapshot()
  })

  it('SEARCH_NASA_MEDIA.SUCCESS', () => {
    expect(medias(undefined, {
      type: SEARCH_NASA_MEDIA.SUCCESS,
      payload: {
        data: [1, 2, 3, 4],
        totalHits: 1234,
        params: {page: 1, q: 'payload'}
      }
    })).toMatchSnapshot()
  })

  it('SEARCH_MORE_NASA_MEDIA.SUCCESS', () => {
    expect(medias({
      data: [66, 55, 35],
      totalHits: 1234,
      params: {page: 1, q: 'payload'},
      loading: true
    }, {
      type: SEARCH_MORE_NASA_MEDIA.SUCCESS,
      payload: {
        data: [1, 2, 3, 4],
        totalHits: 1234,
        params: {page: 1, q: 'payload'}
      }
    })).toMatchSnapshot()
  })

  it('FETCH_MEDIA_ITEM_COLLECTION.SUCCESS', () => {
    expect(medias({
      data: [{}, {}, {}],
      totalHits: 1234,
      params: {page: 1, q: 'payload'},
      loading: true
    }, {
      type: FETCH_MEDIA_ITEM_COLLECTION.SUCCESS,
      payload: {
        thumbUrl: 'thumbUrl-ss',
        mediaUrl: 'mediaUrl-ss'
      },
      meta: {index: 1}
    })).toMatchSnapshot()
  })
})