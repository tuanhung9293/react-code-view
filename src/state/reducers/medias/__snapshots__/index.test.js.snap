// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`medias reducer FETCH_MEDIA_ITEM_COLLECTION.SUCCESS 1`] = `
Object {
  "data": Array [
    Object {},
    Object {
      "mediaUrl": "mediaUrl-ss",
      "thumbUrl": "thumbUrl-ss",
    },
    Object {},
  ],
  "loading": true,
  "params": Object {
    "page": 1,
    "q": "payload",
  },
  "totalHits": 1234,
}
`;

exports[`medias reducer SEARCH_MORE_NASA_MEDIA.PENDING 1`] = `
Object {
  "data": Array [],
  "loading": false,
  "loadingMore": true,
  "params": Object {},
  "totalHits": 0,
}
`;

exports[`medias reducer SEARCH_MORE_NASA_MEDIA.SUCCESS 1`] = `
Object {
  "data": Array [
    66,
    55,
    35,
    1,
    2,
    3,
    4,
  ],
  "loading": true,
  "loadingMore": false,
  "params": Object {
    "page": 1,
    "q": "payload",
  },
  "totalHits": 1234,
}
`;

exports[`medias reducer SEARCH_NASA_MEDIA.PENDING 1`] = `
Object {
  "data": Array [],
  "loading": true,
  "loadingMore": false,
  "params": Object {},
  "totalHits": 0,
}
`;

exports[`medias reducer SEARCH_NASA_MEDIA.SUCCESS 1`] = `
Object {
  "data": Array [
    1,
    2,
    3,
    4,
  ],
  "loading": false,
  "loadingMore": false,
  "params": Object {
    "page": 1,
    "q": "payload",
  },
  "totalHits": 1234,
}
`;

exports[`medias reducer sets initial state 1`] = `
Object {
  "data": Array [],
  "loading": false,
  "loadingMore": false,
  "params": Object {},
  "totalHits": 0,
}
`;
